iana/ip.go

```diff
@@ -0,0 +1,176 @@
+	ipv4Registry []byte
+	// https://www.iana.org/assignments/iana-ipv6-special-registry/iana-ipv6-special-registry.xhtml
+	//go:embed data/iana-ipv6-special-registry-1.csv
+	ipv6Registry []byte
+)
+
+// init parses and loads the embedded IANA special-purpose address registry CSV
+// files for all address families, panicking if any one fails.
+func init() {
+	ipv4Prefixes, err := parseReservedPrefixFile(ipv4Registry, "IPv4")
+	if err != nil {
+		panic(err)
+	}
+
+	ipv6Prefixes, err := parseReservedPrefixFile(ipv6Registry, "IPv6")
+	if err != nil {
+		panic(err)
+	}
+
+	// Add multicast addresses, which aren't in the IANA registries.
+	//
+	// TODO(#8237): Move these entries to IP address blocklists once they're
+	// implemented.
+	additionalPrefixes := []reservedPrefix{
+		{
+			addressFamily: "IPv4",
+			addressBlock:  netip.MustParsePrefix("224.0.0.0/4"),
+			name:          "Multicast Addresses",
+			rfc:           "[RFC3171]",
+		},
+		{
+			addressFamily: "IPv6",
+			addressBlock:  netip.MustParsePrefix("ff00::/8"),
+			name:          "Multicast Addresses",
+			rfc:           "[RFC4291]",
+		},
+	}
+
+	reservedPrefixes = slices.Concat(ipv4Prefixes, ipv6Prefixes, additionalPrefixes)
+
+	// Sort the list of reserved prefixes in descending order of prefix size, so
+	// that checks will match the most-specific reserved prefix first.
+	slices.SortFunc(reservedPrefixes, func(a, b reservedPrefix) int {
+		if a.addressBlock.Bits() == b.addressBlock.Bits() {
+			return 0
+		}
+		if a.addressBlock.Bits() > b.addressBlock.Bits() {
+			return -1
+		}
+		return 1
+	})
+}
+
+// parseReservedPrefixFile parses and returns the IANA special-purpose address
+// registry CSV data for a single address family, or returns an error if parsing
+// fails.
+func parseReservedPrefixFile(registryData []byte, addressFamily string) ([]reservedPrefix, error) {
+	if addressFamily != "IPv4" && addressFamily != "IPv6" {
+		return nil, fmt.Errorf("failed to parse reserved address registry: invalid address family %q", addressFamily)
+	}
+	if registryData == nil {
+		return nil, fmt.Errorf("failed to parse reserved %s address registry: empty", addressFamily)
+	}
+
+	reader := csv.NewReader(bytes.NewReader(registryData))
+
+	// Parse the header row.
+	record, err := reader.Read()
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse reserved %s address registry header: %w", addressFamily, err)
+	}
+	if record[0] != "Address Block" || record[1] != "Name" || record[2] != "RFC" {
+		return nil, fmt.Errorf("failed to parse reserved %s address registry header: must begin with \"Address Block\", \"Name\" and \"RFC\"", addressFamily)
+	}
+
+	// Define regexps we'll use to clean up poorly formatted registry entries.
+	//
+	// 2+ sequential whitespace characters. The csv package takes care of
+	// newlines automatically.
+	whitespacesRE := regexp.MustCompile(`\s{2,}`)
+	// Footnotes at the end, like `[2]`.
+	footnotesRE := regexp.MustCompile(`\[\d+\]$`)
```

Comment:
We typically declare these outside of a function so that they fail at compilation time.

Reply:
Note that this doesn't make a difference in this case.

Putting them outside the function doesn't cause them to fail at compilation time, rather it causes them to fail at package loading time, i.e. startup time. But this function is called from `init()`, so a failure here would _also_ be caught at startup time.